magneticod/magneticod/dht.py

```diff
@@ -189,31 +186,79 @@ def __on_ANNOUNCE_PEER_query(self, message: bencode.KRPCDict, addr: NodeAddress)
         except (TypeError, KeyError, AssertionError):
             return
 
-        self.__outgoing_queue.append((addr, bencode.dumps({
+        data = bencode.dumps({
             b"y": b"r",
             b"t": transaction_id,
             b"r": {
                 b"id": node_id[:15] + self.__true_id[:5]
             }
-        })))
+        })
+        self.sendto(data, addr)
 
         if implied_port:
             peer_addr = (addr[0], addr[1])
         else:
             peer_addr = (addr[0], port)
 
-        self.when_peer_found(info_hash, peer_addr)
+        if not self._is_inforhash_new(info_hash):
+            return
+
+        # create the parent future
+        if info_hash not in self.__tasks:
+            parent_f = self._loop.create_future()
+            parent_f.child_count = 0
+            parent_f.add_done_callback(lambda f: self._parent_task_done(f, info_hash))
+            self.__tasks[info_hash] = parent_f
 
-    def fileno(self) -> int:
-        return self.__socket.fileno()
+        parent_f = self.__tasks[info_hash]
 
-    def __bootstrap(self) -> None:
-        for addr in BOOTSTRAPPING_NODES:
-            self.__outgoing_queue.append((addr, self.__build_FIND_NODE_query(self.__true_id)))
+        if parent_f.done():
+            return
+        if parent_f.child_count > MAX_ACTIVE_PEERS_PER_INFO_HASH:
+            return
+
+        task = asyncio.ensure_future(bittorrent.fetch_metadata(
+            info_hash, peer_addr, self.__max_metadata_size, timeout=PEER_TIMEOUT))
+        task.add_done_callback(lambda task: self._got_child_result(parent_f, task))
+        parent_f.child_count += 1
+        parent_f.add_done_callback(lambda f: task.cancel())
+
+    def _got_child_result(self, parent_task, child_task):
+        parent_task.child_count -= 1
+        try:
+            metadata = child_task.result()
+            if metadata and not parent_task.done():
```

Comment:
Why do we check for `parent_task` being done here when a child got result? I mean, if `parent_task` is done before, and successful, all of its childs will be terminated and this function cannot be called anyway.